File: template/ios/ChatView/Core/Parsing/ChatParsing.swift
// MARK: - ChatParsing.swift
// Парсинг словарей из JS-bridge в доменные модели.

import Foundation

// MARK: - GroupDate formatter (file-private, единственный экземпляр)

private let groupDateFormatter: DateFormatter = {
    let f = DateFormatter()
    f.locale     = Locale(identifier: "en_US_POSIX")
    f.dateFormat = "yyyy-MM-dd"
    return f
}()

// MARK: - ChatMessage + JS parsing

extension ChatMessage {

    /// Создаёт ChatMessage из словаря React Native bridge.
    /// Возвращает nil если отсутствуют обязательные поля (id, timestamp).
    static func from(dict: [String: Any]) -> ChatMessage? {
        guard
            let id          = dict["id"] as? String,
            let timestampMs = dict["timestamp"] as? Double
        else { return nil }

        let timestamp  = Date(timeIntervalSince1970: timestampMs / 1_000)
        let groupDate  = groupDateFormatter.string(from: timestamp)
        let status     = MessageStatus(rawValue: dict["status"] as? String ?? "") ?? .sent
        let isMine     = dict["isMine"]    as? Bool ?? false
        let senderName = dict["senderName"] as? String
        let isEdited   = dict["isEdited"]  as? Bool ?? false

        // Парсинг контента — text, image, video, poll, file
        let textBody  = (dict["text"] as? String).flatMap { $0.isEmpty ? nil : $0 }
        let imageItem = (dict["images"] as? [[String: Any]])?.first.flatMap {
            MessageContent.ImagePayload.from(dict: $0)
        }
        let videoItem = (dict["video"] as? [String: Any]).flatMap {
            MessageContent.VideoPayload.from(dict: $0)
        }
        let pollItem = (dict["poll"] as? [String: Any]).flatMap {
            MessageContent.PollPayload.from(dict: $0)
        }
        let fileItem = (dict["file"] as? [String: Any]).flatMap {
            MessageContent.FilePayload.from(dict: $0)
        }

        // Приоритет: poll > file > video > image > text
        let content: MessageContent
        if let poll = pollItem {
            content = .poll(poll)
        } else if let file = fileItem {
            content = .file(file)
        } else if let video = videoItem {
            if let t = textBody {
                content = .mixedTextVideo(.init(body: t), video)
            } else {
                content = .video(video)
            }
        } else {
            switch (textBody, imageItem) {
            case let (t?, img?): content = .mixed(.init(body: t), img)
            case let (t?, nil):  content = .text(.init(body: t))
            case let (nil, img?): content = .image(img)
            default: return nil
            }
        }

        let reply: ReplyInfo? = (dict["replyTo"] as? [String: Any]).flatMap {
            ReplyInfo.from(dict: $0)
        }

        let actions: [MessageAction] = (dict["actions"] as? [[String: Any]] ?? [])
            .compactMap { MessageAction.from(dict: $0) }

        return ChatMessage(
            id:         id,
            content:    content,
            timestamp:  timestamp,
            senderName: senderName,
            isMine:     isMine,
            groupDate:  groupDate,
            status:     status,
            reply:      reply,
            isEdited:   isEdited,
            actions:    actions
        )
    }
}

// MARK: - MessageContent.ImagePayload + JS parsing

extension MessageContent.ImagePayload {

    static func from(dict: [String: Any]) -> MessageContent.ImagePayload? {
        guard let url = dict["url"] as? String, !url.isEmpty else { return nil }
        return MessageContent.ImagePayload(
            url:          url,
            width:        dict["width"]        as? CGFloat,
            height:       dict["height"]       as? CGFloat,
            thumbnailUrl: dict["thumbnailUrl"] as? String
        )
    }
}

// MARK: - MessageContent.VideoPayload + JS parsing

extension MessageContent.VideoPayload {

    static func from(dict: [String: Any]) -> MessageContent.VideoPayload? {
        guard let url = dict["url"] as? String, !url.isEmpty else { return nil }
        return MessageContent.VideoPayload(
            url:          url,
            thumbnailUrl: dict["thumbnailUrl"] as? String,
            width:        dict["width"]        as? CGFloat,
            height:       dict["height"]       as? CGFloat,
            duration:     dict["duration"]     as? TimeInterval
        )
    }
}

// MARK: - MessageContent.PollPayload + JS parsing

extension MessageContent.PollPayload {

    static func from(dict: [String: Any]) -> MessageContent.PollPayload? {
        guard
            let id       = dict["id"]       as? String,
            let question = dict["question"] as? String,
            let optArr   = dict["options"]  as? [[String: Any]]
        else { return nil }

        let options = optArr.compactMap { MessageContent.PollOption.from(dict: $0) }
        guard !options.isEmpty else { return nil }

        return MessageContent.PollPayload(
            id:               id,
            question:         question,
            options:          options,
            totalVotes:       Int(dict["totalVotes"] as? Double ?? 0),
            selectedOptionId: dict["selectedOptionId"] as? String,
            isClosed:         dict["isClosed"] as? Bool ?? false
        )
    }
}

extension MessageContent.PollOption {

    static func from(dict: [String: Any]) -> MessageContent.PollOption? {
        guard
            let id   = dict["id"]   as? String,
            let text = dict["text"] as? String
        else { return nil }
        return MessageContent.PollOption(
            id:         id,
            text:       text,
            votes:      Int(dict["votes"]      as? Double ?? 0),
            percentage: CGFloat(dict["percentage"] as? Double ?? 0)
        )
    }
}

// MARK: - MessageContent.FilePayload + JS parsing

extension MessageContent.FilePayload {

    static func from(dict: [String: Any]) -> MessageContent.FilePayload? {
        guard
            let url  = dict["url"]  as? String, !url.isEmpty,
            let name = dict["name"] as? String, !name.isEmpty
        else { return nil }
        return MessageContent.FilePayload(
            url:      url,
            name:     name,
            size:     Int64(dict["size"] as? Double ?? 0),
            mimeType: dict["mimeType"] as? String
        )
    }
}

// MARK: - ReplyInfo + JS parsing

extension ReplyInfo {

    static func from(dict: [String: Any]) -> ReplyInfo? {
        guard let id = dict["id"] as? String, !id.isEmpty else { return nil }
        return ReplyInfo(
            replyToId:  id,
            senderName: dict["senderName"] as? String,
            text:       dict["text"]       as? String,
            hasImage:   dict["hasImages"]  as? Bool ?? false
        )
    }
}

// MARK: - MessageAction + JS parsing

extension MessageAction {

    static func from(dict: [String: Any]) -> MessageAction? {
        guard
            let id    = dict["id"]    as? String,
            let title = dict["title"] as? String
        else { return nil }

        return MessageAction(
            id:            id,
            title:         title,
            systemImage:   dict["systemImage"]   as? String,
            isDestructive: dict["isDestructive"] as? Bool ?? false
        )
    }
}

File: template/android/app/src/main/java/com/rnapp/rnchatview/ChatParsing.kt
package com.rnapp.rnchatview

import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

private val groupDateFormatter = SimpleDateFormat("yyyy-MM-dd", Locale.US)

/** Парсит ReadableMap в ChatMessage. Возвращает null если данные невалидны. */
fun ReadableMap.toChatMessage(): ChatMessage? {
    val id = getString("id")?.takeIf { it.isNotEmpty() } ?: return null
    val timestampMs = if (hasKey("timestamp") && !isNull("timestamp"))
        getDouble("timestamp").toLong() else return null

    val groupDate = groupDateFormatter.format(Date(timestampMs))
    val status = MessageStatus.from(if (hasKey("status")) getString("status") else null)
    val isMine = if (hasKey("isMine")) getBoolean("isMine") else false
    val senderName = if (hasKey("senderName") && !isNull("senderName")) getString("senderName") else null
    val isEdited = if (hasKey("isEdited")) getBoolean("isEdited") else false

    val textBody = if (hasKey("text") && !isNull("text"))
        getString("text")?.takeIf { it.isNotEmpty() } else null

    val imagePayload: MessageContent.ImagePayload? =
        if (hasKey("images") && !isNull("images"))
            getArray("images")?.let { arr -> if (arr.size() > 0) arr.getMap(0)?.toImagePayload() else null }
        else null

    val videoPayload: MessageContent.VideoPayload? =
        if (hasKey("video") && !isNull("video")) getMap("video")?.toVideoPayload() else null

    val pollPayload: MessageContent.PollPayload? =
        if (hasKey("poll") && !isNull("poll")) getMap("poll")?.toPollPayload() else null

    val filePayload: MessageContent.FilePayload? =
        if (hasKey("file") && !isNull("file")) getMap("file")?.toFilePayload() else null

    // Приоритет: poll > file > video > image > text
    val content: MessageContent = when {
        pollPayload != null -> MessageContent.Poll(pollPayload)
        filePayload != null -> MessageContent.File(filePayload)
        videoPayload != null -> {
            if (textBody != null)
                MessageContent.MixedTextVideo(MessageContent.TextPayload(textBody), videoPayload)
            else
                MessageContent.Video(videoPayload)
        }
        textBody != null && imagePayload != null ->
            MessageContent.Mixed(MessageContent.TextPayload(textBody), imagePayload)
        textBody != null ->
            MessageContent.Text(MessageContent.TextPayload(textBody))
        imagePayload != null ->
            MessageContent.Image(imagePayload)
        else -> return null
    }

    val reply = if (hasKey("replyTo") && !isNull("replyTo")) getMap("replyTo")?.toReplyInfo() else null

    val actions: List<MessageAction> = if (hasKey("actions") && !isNull("actions"))
        getArray("actions")?.toMessageActions() ?: emptyList()
    else emptyList()

    return ChatMessage(
        id = id,
        content = content,
        timestamp = timestampMs,
        senderName = senderName,
        isMine = isMine,
        groupDate = groupDate,
        status = status,
        reply = reply,
        isEdited = isEdited,
        actions = actions,
    )
}

fun ReadableMap.toMessageAction(): MessageAction? {
    val id = getString("id")?.takeIf { it.isNotEmpty() } ?: return null
    val title = getString("title")?.takeIf { it.isNotEmpty() } ?: return null
    return MessageAction(
        id = id,
        title = title,
        systemImage = if (hasKey("systemImage") && !isNull("systemImage")) getString("systemImage") else null,
        isDestructive = if (hasKey("isDestructive")) getBoolean("isDestructive") else false,
    )
}

fun ReadableMap.toChatInputAction(): ChatInputAction = ChatInputAction.from(
    type = if (hasKey("type")) getString("type") else null,
    messageId = if (hasKey("messageId") && !isNull("messageId")) getString("messageId") else null,
)

fun ReadableArray.toChatMessages(): List<ChatMessage> =
    (0 until size()).mapNotNull { getMap(it)?.toChatMessage() }

fun ReadableArray.toMessageActions(): List<MessageAction> =
    (0 until size()).mapNotNull { getMap(it)?.toMessageAction() }

fun ReadableArray.toEmojiList(): List<String> =
    (0 until size()).mapNotNull { getString(it)?.takeIf { s -> s.isNotEmpty() } }

private fun ReadableMap.toImagePayload(): MessageContent.ImagePayload? {
    val url = getString("url")?.takeIf { it.isNotEmpty() } ?: return null
    return MessageContent.ImagePayload(
        url = url,
        width = if (hasKey("width")) getDouble("width").toFloat() else null,
        height = if (hasKey("height")) getDouble("height").toFloat() else null,
        thumbnailUrl = if (hasKey("thumbnailUrl") && !isNull("thumbnailUrl")) getString("thumbnailUrl") else null,
    )
}

private fun ReadableMap.toVideoPayload(): MessageContent.VideoPayload? {
    val url = getString("url")?.takeIf { it.isNotEmpty() } ?: return null
    return MessageContent.VideoPayload(
        url = url,
        thumbnailUrl = if (hasKey("thumbnailUrl") && !isNull("thumbnailUrl")) getString("thumbnailUrl") else null,
        width = if (hasKey("width")) getDouble("width").toFloat() else null,
        height = if (hasKey("height")) getDouble("height").toFloat() else null,
        duration = if (hasKey("duration")) getDouble("duration") else null,
    )
}

private fun ReadableMap.toPollPayload(): MessageContent.PollPayload? {
    val id = getString("id")?.takeIf { it.isNotEmpty() } ?: return null
    val question = getString("question")?.takeIf { it.isNotEmpty() } ?: return null
    val optArr = if (hasKey("options") && !isNull("options")) getArray("options") else return null
    val options = (0 until optArr!!.size()).mapNotNull { optArr.getMap(it)?.toPollOption() }
    if (options.isEmpty()) return null
    return MessageContent.PollPayload(
        id = id,
        question = question,
        options = options,
        totalVotes = if (hasKey("totalVotes")) getDouble("totalVotes").toInt() else 0,
        selectedOptionId = if (hasKey("selectedOptionId") && !isNull("selectedOptionId")) getString("selectedOptionId") else null,
        isClosed = if (hasKey("isClosed")) getBoolean("isClosed") else false,
    )
}

private fun ReadableMap.toPollOption(): MessageContent.PollOption? {
    val id = getString("id")?.takeIf { it.isNotEmpty() } ?: return null
    val text = getString("text")?.takeIf { it.isNotEmpty() } ?: return null
    return MessageContent.PollOption(
        id = id,
        text = text,
        votes = if (hasKey("votes")) getDouble("votes").toInt() else 0,
        percentage = if (hasKey("percentage")) getDouble("percentage").toFloat() else 0f,
    )
}

private fun ReadableMap.toFilePayload(): MessageContent.FilePayload? {
    val url = getString("url")?.takeIf { it.isNotEmpty() } ?: return null
    val name = getString("name")?.takeIf { it.isNotEmpty() } ?: return null
    return MessageContent.FilePayload(
        url = url,
        name = name,
        size = if (hasKey("size")) getDouble("size").toLong() else 0,
        mimeType = if (hasKey("mimeType") && !isNull("mimeType")) getString("mimeType") else null,
    )
}

private fun ReadableMap.toReplyInfo(): ReplyInfo? {
    val id = getString("id")?.takeIf { it.isNotEmpty() } ?: return null
    return ReplyInfo(
        replyToId = id,
        snapshotSenderName = if (hasKey("senderName") && !isNull("senderName")) getString("senderName") else null,
        snapshotText = if (hasKey("text") && !isNull("text")) getString("text") else null,
        snapshotHasImage = if (hasKey("hasImages")) getBoolean("hasImages") else false,
    )
}
